src/utils.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from functools import wraps
+from typing import Any, Callable, Optional, TypeVar
+
+from ops.charm import CharmBase, EventBase
+from ops.model import BlockedStatus, WaitingStatus
+
+logger = logging.getLogger(__name__)
+
+CharmEventHandler = TypeVar("CharmEventHandler", bound=Callable[..., Any])
+ConditionEvaluation = tuple[bool, str]
+Condition = Callable[[CharmBase], ConditionEvaluation]
+
+
+def container_not_connected(charm: CharmBase) -> ConditionEvaluation:
+    not_connected = not charm._container.can_connect()
+    return not_connected, ("Container is not connected yet" if not_connected else "")
+
+
+def integration_not_exists(integration_name: str) -> Condition:
+    def wrapped(charm: CharmBase) -> ConditionEvaluation:
+        not_exists = not charm.model.relations[integration_name]
+        return not_exists, (f"Missing integration {integration_name}" if not_exists else "")
+
+    return wrapped
+
+
+def block_when(*conditions: Condition) -> Callable[[CharmEventHandler], CharmEventHandler]:
+    def decorator(func: CharmEventHandler) -> CharmEventHandler:
+        @wraps(func)
+        def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
+            event, *_ = args
+            logger.debug(f"Handling event: {event}.")
+
+            for condition in conditions:
+                not_met, msg = condition(charm)
+                if not_met:
+                    event.defer()
+                    charm.unit.status = BlockedStatus(msg)
+                    return None
+
+            return func(charm, *args, **kwargs)
+
+        return wrapper  # type: ignore[return-value]
+
+    return decorator
+
+
+def wait_when(*conditions: Condition) -> Callable[[CharmEventHandler], CharmEventHandler]:
+    def decorator(func: CharmEventHandler) -> CharmEventHandler:
+        @wraps(func)
+        def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
+            event, *_ = args
+            logger.debug(f"Handling event: {event}.")
+
+            for condition in conditions:
+                not_met, msg = condition(charm)
+                if not_met:
+                    event.defer()
+                    charm.unit.status = WaitingStatus(msg)
+                    return None
+
+            return func(charm, *args, **kwargs)
+
+        return wrapper  # type: ignore[return-value]
+
+    return decorator
+
+
+def leader_unit(func: CharmEventHandler) -> CharmEventHandler:
+    @wraps(func)
+    def wrapper(charm: CharmBase, *args: Any, **kwargs: Any) -> Optional[Any]:
+        if not charm.unit.is_leader():
+            return None
+
+        return func(charm, *args, **kwargs)
+
+    return wrapper  # type: ignore[return-value]
```

Comment:
suggestion: We could use https://ops.readthedocs.io/en/latest/index.html#ops.CollectStatusEvent.add_status, these wrappers would act as guards, not calling the function if certain conditions are not met, and the status would be set by another function that would be called after the event has been handled. 

This would make our status handling more consistent, eg think of this case:
```python
@block_when(
    integration_not_exists(KRATOS_INFO_INTEGRATION_NAME),
    integration_not_exists(HYDRA_ENDPOINTS_INTEGRATION_NAME),
)
def handler_a(self): ...

@block_when(
    integration_not_exists(HYDRA_ENDPOINTS_INTEGRATION_NAME),
    integration_not_exists(KRATOS_INFO_INTEGRATION_NAME),
)
def handler_b(self): ...
```

If neither of those integrations are there, when `handler_a` is called the app's status would be `Missing integration with kratos` and when `handler_b` is called the app's status would be `Missing integration with hydra`. Such bugs would be avoided if we handled status logic in a central place, making the decorator's order irrelevant, thus making the code less bug prone

Reply:
👍. I completely forgot this event API (now I remember you mentioned it at some time). I can take a look.